### PR TITLE
Introduce hard ConstantTypes

### PR DESCRIPTION
**Pull request by mbovel**
As other precise types, constant types are currently often widened away. For example:

```scala
val a: 3 = 3
val b /* : Int */ = a
```

For union types, there exists a concept of _softeness_: _hard_ unions are union types explicitly written by the user, while _soft_ unions are created by the compiler (for example as the result of a condition or a match). The former are not widened, while the later are:

```scala
val bool: Boolean = true
val str: String = "hello"
val int: Int = 42
val c /* : Matchable */ = if bool then str else int // Soft union String | Int widened to Matchable
val d: String | Int = if bool then str else int // Explicitly typed as hard union String | Int
val e /* : String | Int */ = d // Hard unions are not widened
```

This PR adds the same _softeness_ concept to constant types, such that constant types explicitly typed by the user are not widened:

```scala
val a: 3 = 3 // Now, this is a hard constant type
val b /* : 3 */ = a // Not widened
```

Previous attempt: https://github.com/lampepfl/dotty/pull/14347, which was only handling singletons inside hard unions.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -3282,7 +3290,12 @@ object Types {
           else tp1.atoms | tp2.atoms
         val tp1w = tp1.widenSingletons
         val tp2w = tp2.widenSingletons
-        myWidened = if ((tp1 eq tp1w) && (tp2 eq tp2w)) this else tp1w | tp2w
+        myWidened =
+          if isSoft then
```
Comment (mbovel) [inline compiler/src/dotty/tools/dotc/core/Types.scala:3294]: With this change, `"a" |<hard> String` would not be widened to `String` anymore. Could add something like `|| tp1w <:< tp2w || tp2w <:< tp1w` here if we want that.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 2]
```diff
@@ -2074,6 +2078,10 @@ object Types {
     }
   }
 
+  trait Softenable {
```
Comment (mbovel) [inline compiler/src/dotty/tools/dotc/core/Types.scala:2081]: Do we need a trait for that? Or could `isSoft` be a method of `Type` directly?

To-do:
- [ ] Find a better name
- [ ] Documentation

**Comment by mbovel** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 2 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/Types.scala:2081]: It's never consumed as a this upper type, and there's no "shared implementation".  So I'd just define/redefine it on SingletonType/ConstantType, and maybe like the docs if they're extensive.  Just IMO :smile:

**Comment by soronpo** [comment]
Could you please describe a little more what this PR is about? 
I may have some relevant input, but I do not know what the scope of this PR is.

**Comment by mbovel** [comment]
> Could you please describe a little more what this PR is about?
> I may have some relevant input, but I do not know what the scope of this PR is.

Description updated 😄 Note: my PRs in draft state are generally just experiments or work in progress and not ready to be reviewed. There are still some tests not passing on this one, but I'd be happy have your input already!

**Comment by soronpo** [comment]
OK, now that I know what this PR is about, here is a feedback regarding the idea (not a review).
To get around widening in Scala is not very simple.
There are several issues. One is regarding what you're trying to solve here.
Another when covariance and contravariance is introduced: https://github.com/lampepfl/dotty/issues/8231
More related discussions:
https://github.com/scala/bug/issues/10838
https://contributors.scala-lang.org/t/singleton-subtyping-and-can-we-mark-scala-singleton-as-experimental/2797

I think there are various ways we've come to live with widening and rely on it without realizing it, so changing this behavior without things breaking is risky business (sadly, because my use-cases suffer from this).

Questions you should be able to answer before preceding: 
1. What do you do with tuples:
```scala
  val x = (2, 2) //is the type (2,2) or (Int, Int)
  val y = (1 | 2, 4) //what is the type here?
  class Foo
  val foo = new Foo
  val z = (1, foo) //is the type (Int, Foo) or (1, foo.type) 
```
2. What do you in if statements:
```scala
  val x = if (cond) 1 else 1 //is the type Int or 1
  val y = if (cond) 1 else 2 //is the type Int or (1 | 2)
```
3. Do you change the widening default when assigning arguments? 
```scala
  def foo[X <: Int](x : X) : X = x
  val x = foo(5) //is the type Int or 5? No need for `& Singleton` ?
```

**Comment by mbovel** [comment]
@soronpo thanks for the pointers! I am starting to see how much various parts of the compiler depend on widening how tricky it is to change it.

This PR tries to address a narrower problem than the general problem of precise types, which is only to preserve constant types explicitly written by the user. For example, we currently have:

```scala
val x: String | Int = ???
val y /* : String | Int */ = x
```

But:

```scala
val x2: "a" | 1 = ???
val y2 /*: Matchable */ = x2
```

Said otherwise, a more precise type is widened to a more general type, which is at they very least counter-intuitive. It seems to me that this could be solved independently of the general problem of precise types.

Note that singleton types in type applications and tuples such are in your first example are preserved:

```scala
val x3: (2, 2) = ???
val y3 /* : (2, 2) */ = x3
```

**Comment by mbovel** [comment]
@soronpo for your examples 2 and 3, it seems like doing less widening in these situations would just break too much existing code. However, I started experiencing with a `precise` modifier that would among other things turn off widening:

```scala
precise def f() =
  val b: Boolean
  val x /* : 1 | 2 = if b then 1 else 2
```

This could be an opt-in way to get more precise types in general and inference in particular. What would you thing about something like that? I'll post my experiment soon so that we can discuss with more material.

**Comment by mbovel** [comment]
In the current state, there at least two problems with this PR:

1. It is assumed in the snippet below (and maybe in other places) that recursively widening a singleton type yields a non-singleton type. This is not true if some singletons are not widened and it results in an infinite loop. https://github.com/lampepfl/dotty/blob/18e3cac34bf397ebb180812b5cb90af2066922f7/compiler/src/dotty/tools/dotc/typer/Implicits.scala#L938-L939
2. `ConstantType`s hashes are used both for hash consing and equality. For the first case, `isSoft` field should be included but not for the later. Doing so breaks atoms comparison for example: https://github.com/lampepfl/dotty/blob/ef25672ed2ce13e6f882f3d59ac8439562458b7c/compiler/src/dotty/tools/dotc/core/TypeComparer.scala#L1426

**Comment by soronpo** [comment]
> @soronpo for your examples 2 and 3, it seems like doing less widening in these situations in general would break too much existing code. However, I started experiencing with a `precise` modifier that would among other things turn off widening:
> 
> ```scala
> precise def f() =
>   val b: Boolean
>   val x /* : 1 | 2 */ = if b then 1 else 2
> ```
> 
> This could be an opt-in way to get more precise types in general and inference in particular. What would you think about something like that? I'll post my early-stage experiment soon so that we can discuss with more material.

I think `exact` instead of `precise` is the better term. I also think it should be a type annotation/modifier and not a term modifier.

**Comment by mbovel** [comment]
- The specific problem of union type loosing their "hardness" described [above](https://github.com/lampepfl/dotty/pull/14360#issuecomment-1027101792) has been fixed by https://github.com/lampepfl/dotty/pull/14399.
- For the question of singletons widening, we need something more general, maybe with an explicit annotation, either at the term or type level. More experiments coming soon.

**Pull request closed without merging**